File: app/models/quizzes/quiz_question/answer_parsers/fill_in_multiple_blanks.rb
#
# Copyright (C) 2013 Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.
#

module Quizzes::QuizQuestion::AnswerParsers
  class FillInMultipleBlanks < AnswerParser
    def parse(question)
      @answers.map_with_group! do |answer_group, answer|
        fields = Quizzes::QuizQuestion::RawFields.new(answer)

        answer = Quizzes::QuizQuestion::AnswerGroup::Answer.new({
          id: fields.fetch_any(:id, nil),
          text: fields.fetch_with_enforced_length([:answer_text, :text]),
          comments: fields.fetch_with_enforced_length([:answer_comment, :comments]),
          weight: fields.fetch_any([:answer_weight, :weight]).to_f,
          blank_id: fields.fetch_with_enforced_length(:blank_id)
        })

        answer_group.taken_ids << answer.set_id(answer_group.taken_ids)
        answer
      end

      question.answers = @answers
      question
    end
  end
end

File: client_apps/canvas_quizzes/apps/events/js/views/question_inspector/answers/fill_in_multiple_blanks.js
/** @jsx React.DOM */
/*
 * Copyright (C) 2014 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

define(function(require) {
  var React = require('old_version_of_react_used_by_canvas_quizzes_client_apps')
  var K = require('../../../constants')
  var NO_ANSWER = require('jsx!./no_answer')
  var keys = Object.keys

  var FIMB = React.createClass({
    statics: {
      questionTypes: [K.Q_FILL_IN_MULTIPLE_BLANKS]
    },

    render: function() {
      var answer = this.props.answer

      return (
        <table>
          {keys(answer).map(function(blank) {
            return (
              <tr key={'blank' + blank}>
                <th scope="row">{blank}</th>
                <td>{answer[blank] || NO_ANSWER}</td>
              </tr>
            )
          })}
        </table>
      )
    }
  })

  return FIMB
})
